import grpc
from concurrent import futures
import time
import sys
import os

# è·¯å¾„ä¿®å¤
sys.path.append(os.path.join(os.path.dirname(__file__), 'proto'))

import ai_service_pb2
import ai_service_pb2_grpc
from core.rag import RAGEngine

# åˆå§‹åŒ–å¼•æ“
engine = RAGEngine()


class AIService(ai_service_pb2_grpc.AIServiceServicer):

    def SyncMessage(self, request, context):
        try:
            user_id = f"User_{request.user_id}"
            # è·å–æ˜µç§°ï¼Œæ²¡æœ‰åˆ™ç”¨ID
            nick = request.nickname if request.nickname else user_id

            engine.add_memory(request.content, user_id, request.msg_id, nick, request.session_id)
            return ai_service_pb2.SyncResponse(code=200)
        except Exception as e:
            print(f"âŒ Error: {e}")
            return ai_service_pb2.SyncResponse(code=500)

    def RevokeMessage(self, request, context):
        try:
            engine.revoke_memory(request.msg_id)
            return ai_service_pb2.RevokeResponse(code=200)
        except Exception as e:
            print(f"âŒ Error Revoke: {e}")
            return ai_service_pb2.RevokeResponse(code=500)

    def SemanticSearch(self, request, context):
        try:
            answer = engine.rag_qa(request.query, request.session_id, request.limit)
            return ai_service_pb2.SearchResponse(code=200, answer=answer)
        except Exception as e:
            return ai_service_pb2.SearchResponse(code=500, answer="æœåŠ¡å¼‚å¸¸")

    def ChatSummary(self, request, context):
        try:
            summary = engine.chat_summary(request.chats)
            return ai_service_pb2.SummaryResponse(code=200, summary=summary)
        except Exception as e:
            return ai_service_pb2.SummaryResponse(code=500)


def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    ai_service_pb2_grpc.add_AIServiceServicer_to_server(AIService(), server)
    server.add_insecure_port('[::]:50051')
    print("ğŸš€ GoNexus-AI æœåŠ¡å·²å¯åŠ¨ (Port: 50051)...", flush=True)
    server.start()
    try:
        while True: time.sleep(86400)
    except KeyboardInterrupt:
        server.stop(0)


if __name__ == '__main__':
    serve()
